lab_3/main.py

```diff
@@ -59,49 +82,93 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
     :param corpus: a tuple of sentences
     :return: a tuple of the encoded sentences
     """
-    pass
+    if not isinstance(storage, LetterStorage) or not isinstance(corpus, tuple):
+        return ()
+    encoded_corpus = []
+    for sentence in corpus:
+        list_of_sentences = []
+        for word in sentence:
+            list_of_sentences.append(tuple([storage.get_id_by_letter(letter) for letter in word]))
+        encoded_corpus.append(tuple(list_of_sentences))
+    return tuple(encoded_corpus)
 
 
 # 6
 class NGramTrie:
 
     def __init__(self, n: int):
-        pass
+        self.size = n
+        self.n_grams = ()
+        self.n_gram_frequencies = {}
+        self.n_gram_log_probabilities = {}
 
     def fill_n_grams(self, encoded_text: tuple) -> int:
         """
         Extracts n-grams from the given sentence, fills the field n_grams
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(encoded_text, tuple):
+            return 1
+        list_of_n_grams = []
+        for sentence in encoded_text:
+            sentence_of_n_grams = []
+            for word in sentence:
+                word_n_grams = []
+                for index in range(len(word) - self.size + 1):
+                    word_n_grams.append(tuple(word[index:index + self.size]))
+                sentence_of_n_grams.append(tuple(word_n_grams))
+            list_of_n_grams.append(tuple(sentence_of_n_grams))
+        self.n_grams = tuple(list_of_n_grams)
+        return 0
 
     def calculate_n_grams_frequencies(self) -> int:
         """
         Fills in the n-gram storage from a sentence, fills the field n_gram_frequencies
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not self.n_grams:
+            return 1
+        for sentence in self.n_grams:
+            for word in sentence:
+                for n_gram in word:
+                    if n_gram in self.n_gram_frequencies:
+                        self.n_gram_frequencies[n_gram] += 1
+                    else:
+                        self.n_gram_frequencies[n_gram] = 1
+        return 0
 
     def calculate_log_probabilities(self) -> int:
         """
         Gets log-probabilities of n-grams, fills the field n_gram_log_probabilities
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not self.n_gram_frequencies:
+            return 1
+        for n_gram in self.n_gram_frequencies:
+            probability_n_gram = self.n_gram_frequencies[n_gram] / \
+                                 sum([self.n_gram_frequencies[another_n_gram] for another_n_gram
+                                      in self.n_gram_frequencies if another_n_gram[0] == n_gram[0]])
```

Comment:
why do we compare 0 indexes?

Reply:
we compare zero indexes because when we sum up frequency we need to find all other tuples with the same fixed first element